observation_portal/requestgroups/dither.py

```diff
@@ -0,0 +1,178 @@
+    '''
+    configuration_dict = dither_details.get('configuration', {})
+    offsets = []
+    instrument_configs = []
+    instrument_config = configuration_dict.get('instrument_configs', [{}])[0]
+    pattern = dither_details.get('pattern')
+    if pattern == 'line':
+        offsets = calc_line_offsets(dither_details.get('num_points'), dither_details.get('point_spacing'), dither_details.get('orientation'),
+                                    dither_details.get('center'))
+    elif pattern == 'box':
+        offsets = calc_box_offsets(dither_details.get('point_spacing'), dither_details.get('line_spacing'),
+                                   dither_details.get('orientation'), sides_angle=270)
+    elif pattern == 'spiral':
+        offsets = calc_spiral_offsets(dither_details.get('num_points'), dither_details.get('point_spacing'), dither_details.get('orientation'))
+    elif pattern == 'grid':
+        offsets = calc_grid_offsets(dither_details.get('num_rows'), dither_details.get('num_columns'), dither_details.get('point_spacing'),
+                                    dither_details.get('line_spacing'), dither_details.get('orientation'), dither_details.get('center'))
+
+
+    for offset in offsets:
+        instrument_config_copy = deepcopy(instrument_config)
+        if 'extra_params' not in instrument_config_copy:
+            instrument_config_copy['extra_params'] = {}
+        instrument_config_copy['extra_params']['offset_ra'] = round(offset[0], 3)
+        instrument_config_copy['extra_params']['offset_dec'] = round(offset[1], 3)
+        instrument_configs.append(instrument_config_copy)
+
+    configuration_dict['instrument_configs'] = instrument_configs
+    return configuration_dict
+
+
+def calc_line_offsets(num_points, point_spacing, orient, center):
+    """Calculate offsets for a LINE dither pattern with <num_points> spaced
+    <point_spacing> arcseconds apart along a line of <orient> degrees East of North
+    (clockwise from North through East)
+    Returns a list of tuples for the offsets"""
+
+    offsets = []
+
+    sino = sin(radians(orient))
+    coso = cos(radians(orient))
+    # A centered line pattern has the midpoint of the line with 0 offset
+    distance_offset = -((num_points-1)*point_spacing) / 2.0 if center else 0.0
+
+    for i in range(0, max(num_points, 0)):
+        distance = i*point_spacing + distance_offset
+        # Angles measured clockwise from North ("y-axis") rather than anti-clockwise
+        # from East ("x-axis") so sin/cos flipped compared to normal
+        x_offset = (distance * sino)
+        y_offset = (distance * coso)
+        offsets.append((x_offset, y_offset))
+    return offsets
+
+
+def normalize_angle(angle):
+    """Normalize an <angle> so that: 0 <= angle_norm < 2 * np.pi"""
+    twopi = 2.0 * pi
+
+    if angle >= twopi:
+        m = floor(angle/twopi)
+        if angle/twopi - m > 0.99999:   # account for rounding errors
```

Comment:
I imagine if your angle was nearly twice 2pi, then m = 1, angle / 2pi < 2, but angle / 2pi - m > 0.99999, which would mean you add one to m, so m=2, and then the norm angle is something slightly less than 2 * 2pi - 2 * 2pi, which would give you a slightly negative angle. That doesn't sound like the intent since the description says normalized angle should be between 0 and 2pi, not negative, so I think this equation isn't great. I'll just remove this since it's only used in the unused box pattern, but for future reference I think a better way to normalize angles would be `angle_norm = angle % (2*pi)`